test/replication-luatest/persistent_gc_test.lua

```diff
@@ -0,0 +1,399 @@
+        t.assert_equals(box.info.gc().consumers[1].vclock, vclock)
+    end, {saved_vclock})
+    cg.master:exec(function()
+        box.error.injection.set('ERRINJ_WAL_GC_PERSIST_FIBER', false)
+    end)
+    -- Check if consumer is advanced after disabling ERRINJ.
+    t.helpers.retrying({}, function()
+        cg.master:exec(function(vclock)
+            vclock[0] = nil
+            t.helpers.retrying({}, function()
+                local consumers = box.space._gc_consumers:select{}
+                t.assert_equals(#consumers, 1)
+                t.assert_equals(consumers[1].vclock, vclock)
+            end)
+        end, {actual_vclock})
+    end)
+end
+
+g.test_gc_consumer_is_not_updated_without_replica = function(cg)
+    cg.master:exec(create_and_fill_space)
+
+    cg.replica:start()
+    cg.replica:wait_for_vclock_of(cg.master)
+    cg.master:exec(function()
+        local consumers = box.space._gc_consumers:select{}
+        t.assert_equals(#consumers, 1)
+        t.assert_equals(consumers[1].vclock[1], box.info.vclock[1])
+    end)
+
+    -- Stop replica and save master vclock and lsn
+    cg.replica:stop()
+    local saved_vclock = cg.master:get_vclock()
+    local saved_lsn = vclock_sum(saved_vclock)
+
+    -- Restart master to check if gc is persisted
+    cg.master:restart()
+    cg.master:exec(load_space)
+    check_xlog(cg.master.workdir, saved_lsn)
+    cg.master:exec(function(vclock)
+        -- Consumer hasn't changed - replica is disabled
+        local consumers = box.space._gc_consumers:select{}
+        t.assert_equals(#consumers, 1)
+        vclock[0] = nil
+        t.assert_equals(consumers[1].vclock, vclock)
+    end, {saved_vclock})
+    check_xlog(cg.master.workdir, saved_lsn)
+end
+
+-- Check if the replica successfully reconnects if its consumer was
+-- manually deactivated
+local function persistent_gc_deactivate_consumer_test(cg, prune_logs)
+    cg.master:exec(create_and_fill_space)
+    local saved_lsn = vclock_sum(cg.master:get_vclock())
+
+    -- Connect replica
+    cg.replica:start()
+    cg.replica:wait_for_vclock_of(cg.master)
+    cg.master:exec(function()
+        local consumers = box.space._gc_consumers:select{}
+        t.assert_equals(#consumers, 1)
+        t.assert_equals(consumers[1].vclock[1], box.info.vclock[1])
+    end)
+
+    -- Stop replica
+    cg.replica:stop()
+
+    cg.master:exec(load_space)
+    check_xlog(cg.master.workdir, saved_lsn)
+    cg.master:exec(function()
+        -- Drop consumer
+        local replica_uuid = box.space._cluster:select{}[2][2]
+        box.space._gc_consumers:replace{replica_uuid}
+        t.assert_equals(box.space._gc_consumers:select(), {{replica_uuid}})
+        t.assert_equals(box.info.gc().consumers, {})
+    end)
+    if prune_logs then
+        -- Trigger snapshot to prune xlogs
+        cg.master:exec(function() box.snapshot() end)
+        check_xlog(cg.master.workdir, vclock_sum(cg.master:get_vclock()))
```

Comment:
```
[001] replication-luatest/persistent_gc_test.lua                      [ fail ]
[001] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/persistent_gc.reject:
[001] Tarantool version is 3.2.0-entrypoint-75-gecbc542e2a
[001] TAP version 13
[001] 1..7
[001] # Started on Thu Jun 13 14:19:33 2024
[001] # Starting group: Indirect replication of _gc_consumers
[001] ok     1	Indirect replication of _gc_consumers.test_indirect_replication_of_consumers
[001] # Starting group: Persistent gc
[001] ok     2	Persistent gc.test_gc_consumer_is_updated_by_relay
[001] ok     3	Persistent gc.test_gc_consumer_update_retry
[001] ok     4	Persistent gc.test_gc_consumer_is_not_updated_without_replica
[001] ok     5	Persistent gc.test_reconnect_inactive_consumer
[001] not ok 6	Persistent gc.test_reconnect_inactive_consumer_and_no_logs
[001] #   ...arantool/test/replication-luatest/persistent_gc_test.lua:43: Assertion failed: 147 > 208
[001] #   stack traceback:
[001] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:43: in function 'check_xlog'
[001] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:269: in function 'persistent_gc_deactivate_consumer_test'
[001] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:283: in function 'Persistent gc.test_reconnect_inactive_consumer_and_no_logs'
[001] #   	...
[001] #   	[C]: in function 'xpcall'
[001] #   artifacts:
[001] #   	master -> /tmp/t/001_replication-luatest/artifacts/master-tG2bbTldx5od
[001] #   	replica -> /tmp/t/001_replication-luatest/artifacts/replica-zwT-oRt_to11
[001] ok     7	Persistent gc.test_deactivate_gc_consumer_for_connected_replica
[001] # Ran 7 tests in 14.327 seconds, 6 succeeded, 1 failed
[001] 
```

Reply:
```
[005] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/persistent_gc.reject:
[005] Tarantool version is 3.2.0-entrypoint-75-gecbc542e2a
[005] TAP version 13
[005] 1..7
[005] # Started on Thu Jun 13 14:19:33 2024
[005] # Starting group: Indirect replication of _gc_consumers
[005] ok     1	Indirect replication of _gc_consumers.test_indirect_replication_of_consumers
[005] # Starting group: Persistent gc
[005] ok     2	Persistent gc.test_gc_consumer_is_updated_by_relay
[005] ok     3	Persistent gc.test_gc_consumer_update_retry
[005] ok     4	Persistent gc.test_gc_consumer_is_not_updated_without_replica
[005] ok     5	Persistent gc.test_reconnect_inactive_consumer
[005] not ok 6	Persistent gc.test_reconnect_inactive_consumer_and_no_logs
[005] #   ...arantool/test/replication-luatest/persistent_gc_test.lua:20: expected: nil, actual: "00000000000000000207.snap"
[005] #   stack traceback:
[005] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:20: in function 'xdir_to_xfiles'
[005] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:29: in function 'check_xlog'
[005] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:269: in function 'persistent_gc_deactivate_consumer_test'
[005] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:283: in function 'Persistent gc.test_reconnect_inactive_consumer_and_no_logs'
[005] #   	...
[005] #   	[C]: in function 'xpcall'
[005] #   artifacts:
[005] #   	master -> Failed to copy artifacts for server (alias: master, workdir: master-nPPw81T5b-GR)
[005] #   	replica -> /tmp/t/005_replication-luatest/artifacts/replica-xERIV_Lg5hqu
[005] ok     7	Persistent gc.test_deactivate_gc_consumer_for_connected_replica
[005] # Ran 7 tests in 14.714 seconds, 6 succeeded, 1 failed
[005] 
```

Reply:
Both these cases flakes a lot on my machine. I build tnt with `cmake -DCMAKE_BUILD_TYPE=Debug -DTEST_BUILD=ON .; make -j`

Reply:
Thanks!
I guess it's because simple replace to a dummy consumer does not help - it's updated back by `persist_fiber` because it is still marked as `async advanced`. I'm working on it now.

Reply:
I guess I've fixed it, but I cannot reproduce your failure. Could you, please, check it again?

Reply:
I cannot see assertion `some number > 208` anymore. But the second error with snap is still there, but backtrace is not as it was:

```
[009] Test failed! Output from reject file /tmp/t/rejects/replication-luatest/persistent_gc.reject:
[009] Tarantool version is 3.2.0-entrypoint-75-g419f5551e4
[009] TAP version 13
[009] 1..7
[009] # Started on Fri Jun 14 14:33:15 2024
[009] # Starting group: Indirect replication of _gc_consumers
[009] not ok 1	Indirect replication of _gc_consumers.test_indirect_replication_of_consumers
[009] #   ...arantool/test/replication-luatest/persistent_gc_test.lua:20: expected: nil, actual: "00000000000000000003.snap"
[009] #   stack traceback:
[009] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:20: in function 'xdir_to_xfiles'
[009] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:63: in function 'check_xlog_le'
[009] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:378: in function 'Indirect replication of _gc_consumers.test_indirect_replication_of_consumers'
[009] #   	...
[009] #   	[C]: in function 'xpcall'
[009] #   artifacts:
[009] #   	server3 -> /tmp/t/009_replication-luatest/artifacts/rs-sudU8BOqMhTu/server3-f-JrLK3t7I2L
[009] #   	server1 -> /tmp/t/009_replication-luatest/artifacts/rs-sudU8BOqMhTu/server1-YsT-SaAu7868
[009] #   	server2 -> /tmp/t/009_replication-luatest/artifacts/rs-sudU8BOqMhTu/server2-lN-oeY406xB6
[009] # Starting group: Persistent gc
[009] ok     2	Persistent gc.test_gc_consumer_is_updated_by_relay
[009] ok     3	Persistent gc.test_gc_consumer_update_retry
[009] ok     4	Persistent gc.test_gc_consumer_is_not_updated_without_replica
[009] ok     5	Persistent gc.test_reconnect_inactive_consumer
[009] ok     6	Persistent gc.test_reconnect_inactive_consumer_and_no_logs
[009] ok     7	Persistent gc.test_deactivate_gc_consumer_for_connected_replica
[009] # Ran 7 tests in 11.267 seconds, 6 succeeded, 1 failed
```

Reply:
> but I cannot reproduce your failure

I build tnt with:

```
cmake -DCMAKE_BUILD_TYPE=Debug -DTEST_BUILD=ON .
```

And run tests with:

```
./test-run.py --builddir=.. --force $(yes persistent_gc | head -100)
```

But I'm afraid, that this flakiness is hardware specific(

Reply:
Thank you! I'll keep looking into it.

Reply:
Do you use Linux on x86?

Reply:
I managed to reproduce these fails on my Linux PC, debugging now.

Reply:
Thanks for pointing out, I've found the reason!
`box.snapshot()` runs cleanup asynchronously - it waits while cleanup is scheduled in WAL thread but not while old files are deleted. That's why there were problems with several snaps and outdated xlogs - they just weren't deleted yet, but were scheduled for deletion. Retries help here.

Reply:
These errors are fixed, thank you

Reply:
Nah, one more place:

```
[007] Tarantool version is 3.2.0-entrypoint-75-g576b00f93a
[007] TAP version 13
[007] 1..7
[007] # Started on Fri Jun 14 18:34:38 2024
[007] # Starting group: Indirect replication of _gc_consumers
[007] ok     1	Indirect replication of _gc_consumers.test_indirect_replication_of_consumers
[007] # Starting group: Persistent gc
[007] ok     2	Persistent gc.test_gc_consumer_is_updated_by_relay
[007] ok     3	Persistent gc.test_gc_consumer_update_retry
[007] ok     4	Persistent gc.test_gc_consumer_is_not_updated_without_replica
[007] ok     5	Persistent gc.test_reconnect_inactive_consumer
[007] not ok 6	Persistent gc.test_reconnect_inactive_consumer_and_no_logs
[007] #   ...arantool/test/replication-luatest/persistent_gc_test.lua:20: expected: nil, actual: "0000
0000000000000198.snap"
[007] #   stack traceback:
[007] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:20: in function 'xdir_to_xfiles'
[007] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:29: in function 'check_xlog'
[007] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:259: in function 'persistent_gc_deactivate_consumer_test'
[007] #   	...arantool/test/replication-luatest/persistent_gc_test.lua:292: in function 'Persistent gc.test_reconnect_inactive_consumer_and_no_logs'
[007] #   	...
[007] #   	[C]: in function 'xpcall'
[007] #   artifacts:
[007] #   	replica -> /tmp/t/007_replication-luatest/artifacts/replica-1tDYxEdMe2kS
[007] #   	master -> /tmp/t/007_replication-luatest/artifacts/master-ThI_alQpxMA7
[007] ok     7	Persistent gc.test_deactivate_gc_consumer_for_connected_replica
[007] # Ran 7 tests in 13.929 seconds, 6 succeeded, 1 failed
```

Maybe retrying inside all `check_xlog_*`?

Reply:
Yep, it seems to be a right approach. Also, revised some moments of the test to make it less flaky.

Reply:
I guess I fixed the test. Even if not, it's not really important - I'll fix it when we'll detect it as a flaky.